src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+-- 'isEmpty'     (overlay x y) == 'isEmpty'   x   && 'isEmpty'   y
+-- 'hasVertex' z (overlay x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (overlay x y) >= 'vertexCount' x
+-- 'vertexCount' (overlay x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (overlay x y) >= 'edgeCount' x
+-- 'edgeCount'   (overlay x y) <= 'edgeCount' x   + 'edgeCount' y
+-- @
+overlay :: (Ord a, Ord b) => AdjacencyMap a b -> AdjacencyMap a b -> AdjacencyMap a b
+overlay (BAM lr1 rl1) (BAM lr2 rl2) = BAM (Map.unionWith Set.union lr1 lr2) (Map.unionWith Set.union rl1 rl2)
+
+-- | /Connect/ two bipartite graphs, not adding the inappropriate edges. This
+-- is an associative operation with the identity 'empty', which distributes
+-- over 'overlay' and obeys the decomposition axion.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory. Note that the
+-- number of edges in the resulting graph is quadratic with respect to the
+-- number of vertices in the arguments: /m = O(m1 + m2 + l1 * r2 + l2 * r1)/.
+--
+-- @
+-- 'isEmpty'     (connect x y) == 'isEmpty'   x   && 'isEmpty'   y
+-- 'hasVertex' z (connect x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (connect x y) >= 'vertexCount' x
+-- 'vertexCount' (connect x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (connect x y) >= 'edgeCount' x
+-- 'edgeCount'   (connect x y) >= 'edgeCount' y
+-- 'edgeCount'   (connect x y) >= 'leftVertexCount' x * 'rightVertexCount' y
+-- 'edgeCount'   (connect x y) <= 'leftVertexCount' x * 'rightVertexCount' y + 'rightVertexCount' x * 'leftVertexCount' y + 'edgeCount' x + 'edgeCount' y
+-- @
+connect :: (Ord a, Ord b) => AdjacencyMap a b -> AdjacencyMap a b -> AdjacencyMap a b
+connect (BAM lr1 rl1) (BAM lr2 rl2) = BAM (Map.unionsWith Set.union [ lr1, lr2,
+                                                                      Map.fromSet (const $ Map.keysSet rl2) (Map.keysSet lr1),
+                                                                      Map.fromSet (const $ Map.keysSet rl1) (Map.keysSet lr2)]) $
+                                           Map.unionsWith Set.union [ rl1, rl2,
+                                                                      Map.fromSet (const $ Map.keysSet lr2) (Map.keysSet rl1),
+                                                                      Map.fromSet (const $ Map.keysSet lr1) (Map.keysSet rl2)]
+
+-- | Constructs the graph comprising two given lists of isolated vertices for
+-- each part.
+-- Complexity: /O(L * log(L))/ time and /O(L)/ memory, where /L/ is the total
+-- length of two lists.
+--
+-- @
+-- vertices [] []                               == 'empty'
+-- vertices [x] []                              == 'leftVertex' x
+-- vertices [] [x]                              == 'rightVertex' x
+-- 'hasLeftVertex' x . ('flip' ('const' [])) vertices == 'elem' x
+-- 'hasRightVertex' x . 'const' [] vertices         == 'elem' x
+-- @
+vertices :: (Ord a, Ord b) => [a] -> [b] -> AdjacencyMap a b
```

Comment:
I think it may be better to follow the approach taken with `vertex`: create `leftVertices`, `rightVertices` and let `vertices` take a list of `Either`s as input. This will make the API closer to non-bipartite graphs.

Reply:
OK, but maybe we would better rename the current version and keep it? Something like `fromPartLists`?
It seems ridiculus to do `overlay (leftVertices ls) (rightVertices rs)` or even `vertices ((Left <$> ls) ++ (Right <$> rs))` in case one needs it. And I needed it in some examples.

Reply:
OK, I agree, let's keep this function as is. Don't worry about adding others for now.